core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -249,7 +249,14 @@ private[spark] class ExecutorAllocationManager(
    * yarn-client mode when AM re-registers after a failure.
    */
   def reset(): Unit = synchronized {
-    initializing = true
+    /**
+     * When some tasks need to be scheduled and initial executor = 0, resetting the initializing
+     * field may cause it to not be set to false in yarn.
+     * SPARK-20079: https://issues.apache.org/jira/browse/SPARK-20079
+     */
+    if (maxNumExecutorsNeeded() == 0) {
+      initializing = true
```

Comment:
@vanzin sorry I think I didn't explain well.

If this flag `initializing` is set to false during initialization, `updateAndSyncNumExecutorsTarget` will recalculate the required executor number and ramp down the executors if there's no job in the current time. And then if first job is submitted, it still requires to ramp up executors to meet the requirement.

For the AM restart scenario I think it is similar during initializing. One exception is the scenario mentioned here, for the case here should ramp up soon to meet the requirement.